validator_test.go

```diff
@@ -0,0 +1,273 @@
+			value: Test{
+				m: map[MessageType]Validator{
+					SimpleEventMessageType: Validators{AlwaysValid},
+				},
+				msg: Message{Type: CreateMessageType},
+			},
+			expectedErr: ErrInvalidMsgType,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.description, func(t *testing.T) {
+			assert := assert.New(t)
+			require := require.New(t)
+			msgv, err := NewTypeValidator(tc.value.m, tc.value.defaultValidators...)
+			require.NotNil(msgv)
+			require.NoError(err)
+			err = msgv.Validate(tc.value.msg)
+			if tc.expectedErr != nil {
+				assert.ErrorIs(err, tc.expectedErr)
+				return
+			}
+
+			assert.NoError(err)
+		})
+	}
+}
+
+func testNewTypeValidator(t *testing.T) {
+	type Test struct {
+		m                 map[MessageType]Validator
+		defaultValidators Validators
+	}
+
+	tests := []struct {
+		description string
+		value       Test
+		expectedErr error
+	}{
+		// Success case
+		{
+			description: "Default Validators success",
+			value: Test{
+				m: map[MessageType]Validator{
+					SimpleEventMessageType: AlwaysValid,
+				},
+				defaultValidators: Validators{AlwaysValid},
+			},
+			expectedErr: nil,
+		},
+		{
+			description: "Empty map of Validators success",
+			value: Test{
+				m:                 map[MessageType]Validator{},
+				defaultValidators: Validators{AlwaysValid},
+			},
+			expectedErr: nil,
+		},
+		{
+			description: "Omit default Validators success",
+			value: Test{
+				m: map[MessageType]Validator{
+					SimpleEventMessageType: Validators{AlwaysValid},
+				},
+			},
+			expectedErr: nil,
+		},
+		// Failure case
+		{
+			description: "Nil list of default Validators error",
+			value: Test{
+				m: map[MessageType]Validator{
+					SimpleEventMessageType: AlwaysValid,
+				},
+				defaultValidators: Validators{nil},
+			},
+			expectedErr: ErrInvalidTypeValidator,
+		},
+		{
+			description: "Empty list of Validators error",
+			value: Test{
+				m: map[MessageType]Validator{
+					SimpleEventMessageType: Validators{},
+				},
+				defaultValidators: Validators{AlwaysValid},
+			},
+			expectedErr: ErrInvalidTypeValidator,
+		},
+		{
+			description: "Nil Validator error",
+			value: Test{
+				m: map[MessageType]Validator{
+					SimpleEventMessageType: nil,
+				},
+				defaultValidators: Validators{AlwaysValid},
+			},
+			expectedErr: ErrInvalidTypeValidator,
+		},
+		{
+			description: "Nil list of Validators error",
+			value: Test{
+				m: map[MessageType]Validator{
+					SimpleEventMessageType: Validators{nil},
+				},
+				defaultValidators: Validators{AlwaysValid},
+			},
+			expectedErr: ErrInvalidTypeValidator,
+		},
+		{
+			description: "Nil map of Validators error",
+			value: Test{
+				m:                 nil,
+				defaultValidators: Validators{AlwaysValid},
+			},
+			expectedErr: ErrInvalidTypeValidator,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.description, func(t *testing.T) {
+			assert := assert.New(t)
+			msgv, err := NewTypeValidator(tc.value.m, tc.value.defaultValidators...)
```

Comment:
How can we verify the `msgv` we've received is useable when the error is nil?

Reply:
Not going to lie, this feels like a trick question 😆
I would think it would be the trust on the api side, but I think I understand where you're coming from in that we should make the api harder to be misused?

In that case we can add an unexported field that is checked during `msgv.Validate(msg)`, where it'll return an automatic error saying `msgv` is bad.

I feel like that should be enough since adding something like `msgv.IsBad()` is kinda redundant since we return an error already `msgv, err := NewTypeValidator(...)` and `msgv.Validate(msg)` can no longer be misuse

Reply:
I was thinking something simpler just in the tests 🙂 we can check and make sure that the map and defaultValidator aren't nil as long as the expectedErr is nil.

Reply:
oooooh, gotcha